### PR TITLE
[UI] Fix block timetable export.

### LINKED ISSUE #58: Downloading attachments not working on Android 11
When attempting to download an attachment from a message on ArrowOS 11.0 (Android 11.0) I'm getting the following error:

`Kod błędu: 50 (on LibrusSandboxDownloadAttachment)
/storage/emulated/0/Szkolny.eu/file.ext: open failed: ENOENT (No such file or directory)`

After manually creating the "Szkolny.eu" folder the error is different:

`Kod błędu: 50 (on LibrusSandboxDownloadAttachment)
/storage/emulated/0/Szkolny.eu/file.ext: open failed: EPERM (Operation not permitted)`

I've already allowed Szkolny.eu to access photos and media in the respective dialog.

After checking the permissions granted to the app in settings it appears as Szkolny.eu doesn't request full storage access which would be needed to create folders and files in the internal storage - I can only switch between "Allow access to media only" and "Deny".

### PR DESCRIPTION
**Pull request by doteq**
Closes #58

### CODE REVIEW

**Comment by kuba2k2** [comment]
Can files be stored under "Szkolny.eu" directory, as it was on previous SDK levels, instead of Pictures?

**Comment by doteq** [comment]
Starting in Android 11 there are two ways to manage files on android. You can use `MANAGE_EXTERNAL_STORAGE` permission, which gives the app access to all data on the device or you can use Scoped Storage. Technically, you can create a directory for your app in external storage - as it was on previous SDK levels, but it will require user interaction - user will be asked to create a directory for your app: 
Screenshot_20210831-160835

The best practice is to store images in the Pictures folder, as described in [here](https://developer.android.com/training/data-storage). However, you can create a subdirectory for your app in the Pictures folder, as I did in the latest commit.

**Comment by kuba2k2** [comment]
Cool, you can also refactor the code a bit not to use duplicate strings and other stuff. The target filename, FOS part could be moved out of the if. The val uri may be set from the return value of if/else.

**Comment by doteq** [comment]
I cleaned up the code a bit and made API levels below 29 to use MediaStore too, which is a much cleaner solution. Now timetable exports to "Szkolny.eu" subdirectory in system Pictures directory on all API levels.

**Comment by kuba2k2** [comment]
Ok, but what about message attachments? Does this work on new API levels, too? Attachments cannot be saved to Pictures directory, so can they be saved elsewhere on new API?

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt [hunk 1]
```diff
@@ -373,25 +376,31 @@ class GenerateBlockTimetableDialog(
...
+            } else {
+                val picturesDirectory = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).absolutePath + File.separator + "Szkolny.eu"
+                File(picturesDirectory).mkdirs()
+                values.put(MediaStore.MediaColumns.DATA, picturesDirectory + File.separator + filename)
+            }
 
             try {
-                val fos = FileOutputStream(outputFile)
+                val uri = resolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values)!!
+                val fos = resolver.openOutputStream(uri)
```
Comment (kuba2k2) [inline app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt:395]: Isn't there a KTX for streams? Something like `resolver.openOutputStream(uri).use { }`.
Also, you're using `!!` too much.

**Review by kuba2k2 — CHANGES_REQUESTED** [review]

**Comment by doteq** [comment]
Oh, I forgot about attachments. They should be saved in the Downloads directory. I'll work on this.

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt (hunk 1 above)
Comment (doteq) [reply app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt]: Something like this?

Comment (kuba2k2) [reply app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt]: I think the idea behind `.use {}` is that you don't have to close it manually.

Comment (doteq) [reply app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt]: Okay, I never used it before.

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt [hunk 2]
```diff
@@ -373,25 +376,31 @@ class GenerateBlockTimetableDialog(
...
+            val filename = "plan_lekcji_${app.profile.name}_${today}_${now}.png"
+            val resolver: ContentResolver = activity.applicationContext.contentResolver
+            val values = ContentValues()
+            values.put(MediaStore.MediaColumns.MIME_TYPE, "image/png")
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                values.put(MediaStore.MediaColumns.DISPLAY_NAME, filename)
+                values.put(MediaStore.MediaColumns.RELATIVE_PATH, Environment.DIRECTORY_PICTURES + File.separator + "Szkolny.eu")
+            } else {
+                val picturesDirectory = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).absolutePath + File.separator + "Szkolny.eu"
```
Comment (kuba2k2) [inline app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt:388]: A cleaner solution would be to work on the File, not on a bare path. The `File()` constructor provides a way to obtain a child file/directory, so you can provide the `Environment...` and "Szkolny.eu" string as its params.

**Review by kuba2k2 — CHANGES_REQUESTED** [review]

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt [hunk 3]
```diff
@@ -373,25 +376,31 @@ class GenerateBlockTimetableDialog(
...
+            val filename = "plan_lekcji_${app.profile.name}_${today}_${now}.png"
+            val resolver: ContentResolver = activity.applicationContext.contentResolver
+            val values = ContentValues()
+            values.put(MediaStore.MediaColumns.MIME_TYPE, "image/png")
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                values.put(MediaStore.MediaColumns.DISPLAY_NAME, filename)
+                values.put(MediaStore.MediaColumns.RELATIVE_PATH, File(Environment.DIRECTORY_PICTURES, "Szkolny.eu").path)
+            } else {
+                val picturesDirectory = File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).absolutePath, "Szkolny.eu")
```
Comment (kuba2k2) [inline app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/timetable/GenerateBlockTimetableDialog.kt:388]: This is still not what I meant.
Take a look at the example code here:
https://developer.android.com/reference/android/os/Environment#getExternalStoragePublicDirectory(java.lang.String)
You don't have to get the absolutePath there. 
Also this method seems to be deprecated for some reason, so you may consider using one of the replacements mentioned in the docs.

**Review by kuba2k2 — CHANGES_REQUESTED** [review]

**Review by kuba2k2 — APPROVED** [review]

**Pull request merged**